# Exploratory Data Analysis - Crime Data Los Angeles

This notebook is used to explore the questions for the Data Scientist Exercise in July 2020.

- How many crimes were reported over the past 5 years?
- List the top five reported crimes for each year for the past 5 years.
- What are the most common MO codes? Have these changed over the past 5 years?
- How else has reported crime changed over time in the City?
- Based on your analysis, please share any changes to services, programs, or policies that the City should consider.

##  Load libraries

In [47]:
# Import libraries
import os, sys, subprocess
import json
import pandas as pd
import numpy as np
from langdetect import detect
import pickle

pd.set_option('display.max_colwidth', -1)
pd.set_option('display.max_columns', 500)

In [24]:
# load project config
terminal_call = ! git rev-parse --show-toplevel
repo_path=terminal_call[0]
project_config_path = os.path.join(repo_path,'project_config.json')

with open(project_config_path,'r') as fp: 
    project_config = json.load(fp)

In [25]:
# import custom module to look at trends
module_path = os.path.join(repo_path,project_config['project_module_relative_path'])
sys.path.append(module_path)

import trends
from trends import get_top_trends as gt
from trends import convert
from trends.convert import crime_dict, mocode_dict # load in crime name dictionary
from importlib import reload # for updating scripts

# Load in data

In [31]:
# load in data that was collected
df = pickle.load(open("../data/crime_data.pkl", "rb"))

In [45]:
# Explore summary info
def eda(df):
    nulls = pd.DataFrame(df.isnull().sum()).T # Check for nulls
    datatypes = pd.DataFrame(df.dtypes).T # Check datatypes
    summary = pd.concat([nulls, datatypes], keys = ["nulls", "datatypes"]) # Create pandas dataframe, because I think it's easier to read
    return summary

In [48]:
eda(df)

dr_no       date_rptd        date_occ time_occ   area area_name  \
nulls     0  0      0               0               0        0      0          
datatypes 0  int64  datetime64[ns]  datetime64[ns]  int64    int64  object     

            rpt_dist_no part_1-2 crm_cd crm_cd_desc mocodes vict_age vict_sex  \
nulls     0  0           0        0      0           240361  0        208478    
datatypes 0  int64       int64    int64  object      object  int64    object    

            vict_descent premis_cd premis_desc weapon_used_cd weapon_desc  \
nulls     0  208527       55        228         1464999        1465000      
datatypes 0  object       float64   object      float64        object       

             status status_desc crm_cd_1 crm_cd_2 crm_cd_3 crm_cd_4 location  \
nulls     0  3       0           11       2063105  2206364  2209991  0         
datatypes 0  object  object      float64  float64  float64  float64  object    

            cross_street      lat      lon   year mocodes_1 mocodes_2  \
nulls     0  1837686      0        0        0      0         240361     
datatypes 0  object       float64  float64  int64  int64     object     

            mocodes_3  
nulls     0  240361    
datatypes 0  object

## How many crimes were reported over the past 5 years?

In [6]:
# Check how many crimes were reported over the past 5 years
# Calculate current date and offset by 5 years
(df["date_rptd"] > (pd.datetime.now()- pd.DateOffset(years=5))).sum()

1117864

### A: 1,117,864 crimes were reported in the past 5 years

## List the top five reported crimes for each year for the past 5 years.

In [ ]:
reload(trends)

### Subset data for the last 5 years

In [33]:
# Select the last 5 years as a dataframe
df_5 = df[df["date_rptd"] > (pd.datetime.now()- pd.DateOffset(years=5))]

In [34]:
# Find the top 5 reported crimes for each year
# Crime is listed in the data under 'crm_cd'
# Custom function 'top_trends' groups by column and selects the top 'n' 
top_crimes = gt.top_trends(df_5, column='year', variable='crm_cd', n=5)

In [35]:
# Add in the name of the crimes to dataframe
top_crimes["crm_name"] = convert.get_names(top_crimes, "crm_cd", crime_dict)

In [36]:
top_crimes

year  crm_cd  count  \
0    2015  624     8853    
1    2015  510     8573    
2    2015  440     7866    
3    2015  330     7316    
4    2015  354     7297    
131  2016  510     18353   
132  2016  624     17942   
133  2016  330     16779   
134  2016  440     14814   
135  2016  310     14558   
261  2017  624     19074   
262  2017  510     18757   
263  2017  330     18082   
264  2017  310     15279   
265  2017  440     14774   
397  2018  624     19427   
398  2018  330     18070   
399  2018  510     17129   
400  2018  440     15420   
401  2018  310     14814   
532  2019  624     19008   
533  2019  330     16750   
534  2019  440     15455   
535  2019  510     15427   
536  2019  740     12859   
668  2020  510     9760    
669  2020  624     8198    
670  2020  330     6893    
671  2020  310     6695    
672  2020  740     6368    

                                                  crm_name  
0    Battery - misdemeanor                                  
1    Stolen Vehicle                                         
2    Theft - $950 & under                                   
3    Burg from Vehicle                                      
4    Theft of Identity                                      
131  Stolen Vehicle                                         
132  Battery - misdemeanor                                  
133  Burg from Vehicle                                      
134  Theft - $950 & under                                   
135  Burglary                                               
261  Battery - misdemeanor                                  
262  Stolen Vehicle                                         
263  Burg from Vehicle                                      
264  Burglary                                               
265  Theft - $950 & under                                   
397  Battery - misdemeanor                                  
398  Burg from Vehicle                                      
399  Stolen Vehicle                                         
400  Theft - $950 & under                                   
401  Burglary                                               
532  Battery - misdemeanor                                  
533  Burg from Vehicle                                      
534  Theft - $950 & under                                   
535  Stolen Vehicle                                         
536  Vandalism - Felony $400 & over, all church vandalisms  
668  Stolen Vehicle                                         
669  Battery - misdemeanor                                  
670  Burg from Vehicle                                      
671  Burglary                                               
672  Vandalism - Felony $400 & over, all church vandalisms

## What are the most common MO codes? Have these changed over the past 5 years?

From the summary info above, I see there are 240361 rows of missing MO codes in the original dataframe. I will remove the NA for this question.

In [49]:
df["mocodes"].notnull()

0        True 
1        True 
2        True 
3        True 
4        True 
         ...  
95400    False
95401    True 
95402    True 
95403    True 
95404    True 
Name: mocodes, Length: 2210104, dtype: bool

In [52]:
# Select the last 5 years as a dataframe
df_5 = df_5[df_5["mocodes"].notnull()]

In [53]:
# Find the top 10 most common MO codes across all the years (2010-2020)
df_5["mocodes"].value_counts(ascending = False).head(10)

0344         93842
0329         44407
1501         18472
0325         12737
0416         12317
0329 1300    8302 
1822         8234 
0344 1300    6090 
0344 1606    5898 
0329 1307    4808 
Name: mocodes, dtype: int64

In [54]:
# Find the top 10 MO codes for each year
# MO code is listed in the data under 'mocodes'
# Custom function 'top_trends' groups by column and selects the top 'n' 
top_mo = gt.top_trends(df_5, column='year', variable='mocodes_1', n=10)

In [55]:
# Add in the name of the MO codes to dataframe
top_mo["mo_name"] = convert.get_names(top_mo, "mocodes_1", mocode_dict)

In [58]:
top_mo

year  mocodes_1  count                      mo_name
0     2015  344        24853  Removes vict property      
1     2015  329        9307   Vandalized                 
2     2015  2000       5622   Domestic violence          
3     2015  416        5152   Hit-Hit w/ weapon          
4     2015  325        2820   Took merchandise           
5     2015  1822       2775   Stranger                   
6     2015  400        2727   Force used                 
7     2015  1501       2473   Other MO (see rpt)         
8     2015  100        2095   Suspect Impersonate        
9     2015  1402       1774   Evidence Booked (any crime)
412   2016  344        47959  Removes vict property      
413   2016  329        19660  Vandalized                 
414   2016  2000       11630  Domestic violence          
415   2016  416        10266  Hit-Hit w/ weapon          
416   2016  1822       6693   Stranger                   
417   2016  325        5825   Took merchandise           
418   2016  400        5682   Force used                 
419   2016  1501       5419   Other MO (see rpt)         
420   2016  1402       3803   Evidence Booked (any crime)
421   2016  100        3660   Suspect Impersonate        
890   2017  344        46421  Removes vict property      
891   2017  329        19909  Vandalized                 
892   2017  2000       11675  Domestic violence          
893   2017  416        10944  Hit-Hit w/ weapon          
894   2017  1822       6799   Stranger                   
895   2017  1501       5658   Other MO (see rpt)         
896   2017  400        5600   Force used                 
897   2017  325        5297   Took merchandise           
898   2017  1609       4051   Smashed                    
899   2017  1402       3892   Evidence Booked (any crime)
1395  2018  344        44775  Removes vict property      
1396  2018  329        18866  Vandalized                 
1397  2018  2000       10744  Domestic violence          
1398  2018  416        10138  Hit-Hit w/ weapon          
1399  2018  1822       8016   Stranger                   
1400  2018  400        6796   Force used                 
1401  2018  325        5227   Took merchandise           
1402  2018  1501       4731   Other MO (see rpt)         
1403  2018  1300       4666   Vehicle involved           
1404  2018  1609       4044   Smashed                    
1917  2019  344        41352  Removes vict property      
1918  2019  329        17870  Vandalized                 
1919  2019  2000       10275  Domestic violence          
1920  2019  416        8893   Hit-Hit w/ weapon          
1921  2019  1822       8773   Stranger                   
1922  2019  400        6929   Force used                 
1923  2019  1300       5482   Vehicle involved           
1924  2019  325        5187   Took merchandise           
1925  2019  1501       4017   Other MO (see rpt)         
1926  2019  1402       3851   Evidence Booked (any crime)
2434  2020  344        15521  Removes vict property      
2435  2020  1822       8101   Stranger                   
2436  2020  329        7932   Vandalized                 
2437  2020  913        4498   Victim knew Suspect        
2438  2020  2000       4184   Domestic violence          
2439  2020  416        3133   Hit-Hit w/ weapon          
2440  2020  1300       3091   Vehicle involved           
2441  2020  400        2544   Force used                 
2442  2020  1402       1952   Evidence Booked (any crime)
2443  2020  325        1725   Took merchandise